core/src/ast/macros.rs

```diff
@@ -114,43 +153,342 @@ impl Parse for MacroIdent {
         let ident: Ident = input.parse()?;
         input.parse::<Token![:]>()?;
         let ty: Ident = input.parse()?;
-        Ok(Self { ident, _ty: ty })
+        Ok(Self { ident, ty })
+    }
+}
+
+/// Hack to read information from the TokenTree while saving the rest to be later copied into a buffer.
+/// Will attempt to read `T`, and on a success all of the unread tokens after the stream will be written to `remaining`.
+struct MaybeParse<T: Parse> {
+    item: T,
+    remaining: TokenStream,
+}
+
+impl<T: Parse> Parse for MaybeParse<T> {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let item = input.parse::<T>()?;
+        let remaining = input.parse()?;
+        Ok(Self { item, remaining })
+    }
+}
+
+impl<T: Parse> MaybeParse<T>
+where
+    MacroFrag: From<T>,
+{
+    fn try_parse(
+        i: &MacroIdent,
+        cursor: &Cursor,
+        args: &mut HashMap<Ident, MacroFrag>,
+    ) -> syn::Result<TokenBuffer> {
+        let out = syn::parse2::<MaybeParse<T>>(cursor.token_stream())?;
+
+        args.insert(i.ident.clone(), MacroFrag::from(out.item));
+
+        let buf = TokenBuffer::new2(out.remaining);
+        Ok(buf)
     }
 }
 
 #[derive(Debug)]
 #[non_exhaustive]
-/// Struct for defining a macro (i.e., `macro_rules! example`)
-pub struct MacroDef {
-    pub match_tokens: HashMap<Ident, usize>,
-    pub body: TokenStream,
+pub struct MacroUse {
+    args: HashMap<Ident, MacroFrag>,
 }
 
-impl Parse for MacroDef {
+impl MacroUse {
+    /// Not an official implementation of [`syn::parse::Parse`], since we need to know the [`MacroDef`] to know how to extract [`MacroUse::args`].
+    fn parse(def: &MacroDef, stream: TokenStream) -> syn::Result<Self> {
+        let mut args = HashMap::new();
+
+        Self::parse_macro_matcher(&def.matcher.matches, stream, &mut args)?;
+
+        Ok(Self { args })
+    }
+
+    fn parse_macro_matcher(
+        matches: &[MacroMatch],
+        stream: TokenStream,
+        args: &mut HashMap<Ident, MacroFrag>,
+    ) -> syn::Result<()> {
+        let mut buf = TokenBuffer::new2(stream);
+        let mut c = buf.begin();
+
+        let mut match_iter = matches.iter();
+        while let Some((tt, next)) = c.token_tree() {
+            let curr_match = match_iter.next();
+            match curr_match {
+                Some(MacroMatch::Ident(i)) => match i.ty.to_string().as_str() {
```

Comment:
suggestion: this can probably be turned into a method on MacroMatch, like `consume_tokens()`